README.md
# kdp-report

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-12-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-12-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/kdp-report/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/kdp-report/compare/0.1.0...0.1.1


test/kdp_report/core_test.clj
(ns kdp-report.core-test
  (:require [clojure.test :refer :all]
            [kdp-report.kdp :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/kdp_report/d2d.clj
(ns kdp-report.d2d
  (:require [clojure.data.csv :as csv]
            [clojure.java.io :as io]
            [clojure.pprint :refer :all]))

; go to statements -> downloads and download each 'sales report'
; or just do raw sales data by month

(def d2d-filenames
  ["/Users/adrian.osullivan/Dropbox/kdp/2016-01-rawdata.csv"
   "/Users/adrian.osullivan/Dropbox/kdp/2016-02-rawdata.csv"
   "/Users/adrian.osullivan/Dropbox/kdp/2016-03-rawdata.csv"
   "/Users/adrian.osullivan/Dropbox/kdp/2016-04-rawdata.csv"
   "/Users/adrian.osullivan/Dropbox/kdp/2016-05-rawdata.csv"
   "/Users/adrian.osullivan/Dropbox/kdp/2016-06-rawdata.csv"])

(defrecord Row
  [StartDate	EndDate	Publisher	TitleID	ISBN	Title	PrimaryAuthor	Distributor	Vendor	Country	UnitsSold	UnitsReturned	NetUnitSales	ListPricePerUnit	CurrencyCode	OfferPricePerUnit	RoyaltyPercent	SalesChannelFeeTaxesPerUnit	SalesChannelRevenuePerUnit	ExtendedSalesChannelRevenue	SalesChannelShare	D2DShare	PublisherShare	PublisherShareUSDestimated	Verified])

(defn read-rows [filename]
  (->>
    (with-open [in-file (io/reader filename)]
      (doall (csv/read-csv in-file)))
    (rest)))

(defn books-ext [filename]
  (->>
    (for [row (read-rows filename)]
      (apply ->Row (flatten row)))
    (map #(into {} %))))

(defn books-all [filename]
  (map (fn [book] (select-keys book [:Title :UnitsSold :CurrencyCode :Country :Vendor]))
    (books-ext filename)))

(defn parse-int [s]
  (Integer/parseInt (re-find #"\A-?\d+" s)))

(defn expand-units-sold [books]
  (map (fn [book] (select-keys book [:Title :CurrencyCode :Country :Vendor]))
     (flatten
       (for [book books]
         (let [n (parse-int (:UnitsSold book))]
           (if (> n  1)  (repeat n book) book))))))

(def d2d-books
  (->>
    (map books-all d2d-filenames)
    (map expand-units-sold )
    (flatten)))







src/kdp_report/core.clj
(ns kdp-report.core
  (:require [dk.ative.docjure.spreadsheet :refer :all]
            [clojure.pprint :refer :all] ))


(def filenames
  ["/Users/adrian.osullivan/Dropbox/kdp/kdp-report-1-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-2-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-3-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-4-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-5-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-6-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-7-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-reports-08-2016-1479749593121-efe96ea153478e0814a09f54e3184f79.xlsx"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-reports-09-2016-1479749586926-f88c1bff260960ac54d1c5e30e59c834.xlsx"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-reports-10-2016-1479748641878-49ecf7cf74852ad18001bc12d9788161.xlsx"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-reports-11-2016-1482242511043-699037402df131d60bfb9ef6023b6d9c.xlsx"
   ])


;date - determined by the filename


(def currencies {"Amazon Kindle US Store" "USD"
                 "Amazon Kindle UK Store" "GBP"
                 "Amazon Kindle DE Store" "EUR (Germany)"
                 "Amazon Kindle FR Store" "EUR (France)"
                 "Amazon Kindle Japan Store" "JPY"
                 "Amazon Kindle CA Store" "CAD"
                 "Amazon Kindle IT Store" "EUR (Italy)"
                 "Amazon Kindle ES Store" "EUR (Spain)"
                 "Amazon Kindle IN Store" "INR"
                 "Amazon Kindle AU Store" "AUD"
                 "Amazon Kindle NL Store" "EUR (Netherlands)"
                 "Amazon Kindle BR Store" "BRL"
                 "Amazon Kindle MX Store" "MXN"})

(defn is-royalty-row? [rrow]
  (boolean (re-find #"Total Royalty" (rrow :row-desc))))

(defn is-header-row? [rrow]
  (= "Title" (rrow :row-desc)))

(defn is-nosales-row? [rrow]
  (= "There were no sales during this period" (rrow :row-desc)))

(defn is-report-row? [rrow]
  (boolean
    (re-find #"Sales report for the period" (rrow :row-desc))))

(defn is-currency-row? [rrow]
  (nil? (rrow :row-desc)))

(defn ws [filename]
  (->> (load-workbook filename)
       (select-sheet
         (if (.endsWith filename ".xls") "Reports" "eBook Royalty Report"))))

(defn ws-seq [filename]
  (drop-last 6
    (remove is-report-row?
      (remove is-header-row?
        (remove is-royalty-row?
          (remove is-currency-row?
            (filter #(some? (%1 :row-desc)) ;filter out if nothing in first col
               (remove nil?
                 (select-columns
                   {:A :row-desc, :C :asin,  :D :units, :G :royalty, :I :list-price, :M :royalty-paid} (ws filename))))))))))
s

(def section?
  (comp #(contains? currencies %) :row-desc))

(defn trans-rows [filename]
  (map flatten
     (partition 2
       (partition-by  section? (ws-seq filename)))))

(defn replace-first-with-currency [row]
  (assoc row 0 (currencies (:row-desc (first row)))) )

(defn curr-books-map [filename]
  (->> (map #(into [] %) (trans-rows filename))
       (map replace-first-with-currency)))

(defn get-all [filename]
  ;(filter #(= "Rain of Clarity" (second %))
  (remove #(= "There were no sales during this period" (second %))
    (remove #(= "" (second %))
      (partition 2
        (flatten
          (for [row (curr-books-map filename)]
            (for [book (rest row)]
              [(first row) (book :row-desc)])))))))

(def all-kdp (partition 2
                        (flatten
                          (map get-all filenames))))

(group-by second all-kdp)


; -----------------

(get-all "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-1-2016.xls")



(+ 1 2)



src/kdp_report/kdp.clj
(ns kdp-report.kdp
  (:require [dk.ative.docjure.spreadsheet :refer :all]))

;download 'prior months royalties one by one'

(def kdp-filenames
  ["/Users/adrian.osullivan/Dropbox/kdp/kdp-report-1-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-2-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-3-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-4-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-5-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-6-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-report-7-2016.xls"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-reports-08-2016-1479749593121-efe96ea153478e0814a09f54e3184f79.xlsx"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-reports-09-2016-1479749586926-f88c1bff260960ac54d1c5e30e59c834.xlsx"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-reports-10-2016-1479748641878-49ecf7cf74852ad18001bc12d9788161.xlsx"
   "/Users/adrian.osullivan/Dropbox/kdp/kdp-reports-11-2016-1482242511043-699037402df131d60bfb9ef6023b6d9c.xlsx"])

(def currencies {"Amazon Kindle US Store" "USD"
                 "Amazon Kindle UK Store" "GBP"
                 "Amazon Kindle DE Store" "EUR (Germany)"
                 "Amazon Kindle FR Store" "EUR (France)"
                 "Amazon Kindle Japan Store" "JPY"
                 "Amazon Kindle CA Store" "CAD"
                 "Amazon Kindle IT Store" "EUR (Italy)"
                 "Amazon Kindle ES Store" "EUR (Spain)"
                 "Amazon Kindle IN Store" "INR"
                 "Amazon Kindle AU Store" "AUD"
                 "Amazon Kindle NL Store" "EUR (Netherlands)"
                 "Amazon Kindle BR Store" "BRL"
                 "Amazon Kindle MX Store" "MXN"})

(def currency-maps {"USD" {:CurrencyCode "USD", :Country "US"}
                    "GBP" {:CurrencyCode "GBP", :Country "GB"}
                    "EUR (Germany)" {:CurrencyCode "EUR", :Country "DE"}
                    "EUR (France)" {:CurrencyCode "EUR", :Country "FR"}
                    "JPY" {:CurrencyCode "JPY", :Country "JP"}
                    "CAD" {:CurrencyCode "CAD", :Country "CA"}
                    "EUR (Italy)" {:CurrencyCode "EUR", :Country "IT"}
                    "EUR (Spain)" {:CurrencyCode "EUR", :Country "ES"}
                    "INR" {:CurrencyCode "INR", :Country "IN"}
                    "AUD" {:CurrencyCode "AUD", :Country "AU"}
                    "EUR (Netherlands)" {:CurrencyCode "EUR", :Country "NE"}
                    "BRL" {:CurrencyCode "BRL", :Country "BR"}
                    "MXN" {:CurrencyCode "MXN", :Country "MX"}})

(defn is-royalty-row? [rrow]
  (boolean (re-find #"Total Royalty" (rrow :row-desc))))

(defn is-header-row? [rrow]
  (= "Title" (rrow :row-desc)))

(defn is-nosales-row? [rrow]
  (= "There were no sales during this period" (rrow :row-desc)))

(defn is-report-row? [rrow]
  (boolean
    (re-find #"Sales report for the period" (rrow :row-desc))))

(defn is-currency-row? [rrow]
  (nil? (rrow :row-desc)))

(defn ws [filename]
  (->> (load-workbook filename)
       (select-sheet
         (if (.endsWith filename ".xls") "Reports" "eBook Royalty Report"))))

(defn ws-seq [filename]
  (->>
    (ws filename)
    (select-columns
      {:A :row-desc, :C :asin,  :D :units, :G :royalty, :I :list-price, :M :royalty-paid})
    (remove nil?)
    (filter #(some? (%1 :row-desc)))
    (remove is-report-row?)
    (remove is-header-row?)
    (remove is-royalty-row?)
    (remove is-currency-row?)
    (drop-last 6)))

(def section?
  (comp #(contains? currencies %) :row-desc))

(defn trans-rows [filename]
  (map flatten
     (partition 2
       (partition-by section? (ws-seq filename)))))

(defn replace-first-with-currency [row]
  (assoc row 0 (currencies (:row-desc (first row)))) )

(defn curr-books-map [filename]
  (->> (map #(into [] %) (trans-rows filename))
       (map replace-first-with-currency)))

(defn get-all [filename]
  (->>
    (for [row (curr-books-map filename)
          book (rest row)]
      [(first row) (book :row-desc)])
    (flatten)
    (partition 2)
    (remove #(= "" (second %)))
    (remove #(= "There were no sales during this period" (second %)))))
    ;(filter #(= "Rain of Clarity" (second %))

(def all-kdp
  (partition 2
    (flatten
      (map get-all kdp-filenames))))

(defn to-book-map [[curr title]]
  (assoc {} :Title title
            :CurrencyCode (:CurrencyCode (currency-maps curr))
            :Country (:Country (currency-maps curr))
            :Vendor "Amazon.com"))

(def kindle-books
  (->>
    (map vec all-kdp)
    (map to-book-map)
    (flatten)))
